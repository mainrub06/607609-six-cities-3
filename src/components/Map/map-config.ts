export const mapConfig = {
  center: [52.38333, 4.9] as [number, number],
  zoom: 12 as number,
  zoomControl: false as boolean,
  marker: true as boolean,
  icons: {
    blue: {
      iconUrl: `/img/pin.svg` as string,
      iconSize: [30, 45] as [number, number],
    },
    orange: {
      iconUrl: `/img/pin-active.svg` as string,
      iconSize: [30, 45] as [number, number],
    },
  },
};

export const titleLayerConfig = {
  options: {
    attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`,
  },
  urlTemplate: `https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`,
};
