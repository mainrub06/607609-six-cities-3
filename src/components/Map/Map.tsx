import React, { PureComponent, createRef, RefObject } from "react";
import leaflet from "leaflet";
import { IOffer, IPoint } from "../../types/offer";
import { mapConfig, titleLayerConfig } from "./map-config";

interface IMapState {
  city: [number, number];
  zoom: number;
  mapConfig: {
    center: [number, number];
    zoom: number;
    zoomControl: boolean;
    marker: boolean;
  };
}

interface IMapProps {
  nearMap: boolean;
  offers: IOffer[];
  activeOfferId: string;
  city: {
    name: string;
    location: IPoint;
  };
}

class Map extends PureComponent<IMapProps, IMapState> {
  mainMapRef: RefObject<HTMLElement>;
  icons: any;
  map: null | any;
  layerGroupStorage: null | any;

  constructor(props: IMapProps) {
    super(props);

    this.icons = {
      iconBlue: leaflet.icon(mapConfig.icons.blue),
      iconOrange: leaflet.icon(mapConfig.icons.orange),
    };

    this.mainMapRef = createRef();
    this.map = null;
    this.layerGroupStorage = null;
  }

  componentDidMount() {
    const { city } = this.props;
    const cityLocation = [city.location.latitude, city.location.longitude];

    if (this.mainMapRef.current) {
      this.map = leaflet.map(this.mainMapRef.current, mapConfig);
      this.map.setView(cityLocation, city.location.zoom);

      leaflet
        .tileLayer(titleLayerConfig.urlTemplate, titleLayerConfig.options)
        .addTo(this.map);

      this.layerGroupStorage = {
        map: this.map,
        layerGroup: leaflet.layerGroup().addTo(this.map),
      };

      this.updateMap();
    }
  }

  componentDidUpdate(prevState: any) {
    if (
      prevState.offers !== this.props.offers ||
      prevState.activeOfferId !== this.props.activeOfferId
    ) {
      const { layerGroup } = this.layerGroupStorage;
      const { city } = this.props;
      const cityLocation = [city.location.latitude, city.location.longitude];

      this.map.setView(cityLocation, city.location.zoom);

      layerGroup.clearLayers();
      this.updateMap();
    }
  }

  componentWillUnmount() {
    this.map = null;
  }

  updateMap() {
    const { offers, activeOfferId } = this.props;
    const { layerGroup } = this.layerGroupStorage;

    offers.forEach((offer) => {
      const icon =
        activeOfferId && activeOfferId === offer.id
          ? this.icons.iconOrange
          : this.icons.iconBlue;

      leaflet.marker(offer.cords, { icon }).addTo(layerGroup);
    });
  }

  render() {
    return (
      <section
        ref={this.mainMapRef}
        className={`${
          this.props.nearMap ? `property__map` : `cities__map`
        } map`}
      />
    );
  }
}

export default Map;
