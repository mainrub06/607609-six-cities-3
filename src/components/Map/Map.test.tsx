import React from "react";
import renderer from "react-test-renderer";
import Map from "./Map";
import PLACE_MOCK from "../../mock/offers";

const CITY_MOCK = {
  name: `Paris`,
  location: {
    latitude: 1.1,
    longitude: 1.2,
    zoom: 13,
  },
};

it(`Render snapshot <Map/>`, () => {
  const tree = renderer
    .create(
      <Map
        city={CITY_MOCK}
        activeOfferId={PLACE_MOCK[0].id}
        offers={PLACE_MOCK}
        nearMap={true}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
