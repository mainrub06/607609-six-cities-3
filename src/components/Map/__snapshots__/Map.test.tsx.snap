// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render snapshot <Map/> 1`] = `
<section
  className="property__map map"
/>
`;
